### DISCUSSION TITLE
Cannot access State in a FromRequestParts implementation

### DISCUSSION THREAD

**Discussion by jkobz** [tokio-rs/axum#2763] [Q&A] [answered]
### Summary

When I attempt to extract value in a handler like so...

#### handler.rs
```
pub async fn sign_in(
    custom::extract::User(user): custom::extract::User,
```

...error below is raised:

> the trait bound `state::State: FromRef<()>` is not satisfied.

My initial attempt to implement `FromRequestParts`:

#### extract.rs

```
    #[axum::async_trait]
    impl axum::extract::FromRequestParts<custom::state::State> for User {
        type Rejection = custom::error::Error;

        async fn from_request_parts(
            mut parts: &mut axum::http::request::Parts,
            state: &custom::state::State,
        ) -> Result<Self, Self::Rejection> {
```

After some back-and-forth I came up with the second attempt to no avail:

#### extract.rs
```
    #[axum::async_trait]
    impl<S> axum::extract::FromRequestParts<S> for User
    where
        custom::state::State: axum::extract::FromRef<S>,
        S: Send + Sync,
    {
        type Rejection = custom::error::Error;

        async fn from_request_parts(
            mut parts: &mut axum::http::request::Parts,
            state: &S,
        ) -> Result<Self, Self::Rejection> {
            let state = parts
                .extract_with_state::<axum::extract::State<modules::state::State>>(state)
                .await?;
```

I feel like my error is unrelated to `FromRequestParts` implementation, maybe there is something wrong with my `Router` setup.

#### main.rs

```
let app = NormalizePathLayer::trim_trailing_slash().layer(
            routes::router(&self.assets_path)
                .with_state(state)
                .layer(middlewares),
        );
```

#### routes.rs

```
pub fn router<'a>(assets_path: &'a str) -> axum::Router<custom::state::State> {
    axum::Router::new().nest("/auth", auth::router())
```

#### auth.rs

```
pub fn router() -> axum::Router<custom::state::State> {
    axum::Router::new()
        .route("/", axum::routing::post(handler::sign_in)) // See handler.rs snippet above
```

### axum version

0.7.5

**Comment by jkobz** [comment]
Btw my state is:

```
#[derive(Clone, Debug)]
pub struct State(pub Arc<Inner>);
```

**Comment by jplatte** [comment, marked as answer]
Could it be that you're using `#[debug_handler]` on your handler function without specifying the state type? It will only be inferred if you use the `State` extractor.

**Reply by jkobz** [reply]
> Could it be that you're using `#[debug_handler]` on your handler function without specifying the state type? It will only be inferred if you use the `State` extractor.

Yeah, indeed. For the future generations: have a look at [this debug_handler](https://docs.rs/axum/0.7.5/axum/attr.debug_handler.html#changing-state-type) section.
